bv_to_int preprocessing pass

yoni206 opened the pull request:
This PR introduces a new preprocessing pass, called bv_to_int.
The purpose of this pass is to translate bit-vector formulas to integer formulas, and then use an arithmetic solver rather than the bit-vector solver.
This is the first PR in this direction, which offers the most basic translation from bit-vectors to integers.
Some simple optimizations are already included in this PR, thanks to @ahmed-irfan .
Future pull requests are planned to improve and further optimize the translation.

yoni206 commented:
Since I am having bugs on SMT-LIB benchmarks, I am closing the PR until it becomes stable.

yoni206 commented:
All issues from SMT-LIB benchmarks are now fixed, so I am re-opening the PR.

yoni206 commented:
Actually, some merges with current master need to be done before reopening (specifically with setting the right logic, after the Ackermannization changes).

yoni206 commented:
All issues from merging with master are now fixed, so I am re-opening the PR.

yoni206 commented on src/preprocessing/passes/bv_to_int.cpp:
```diff
@@ -0,0 +1,1003 @@
+                Node zero = d_nm->mkConst<Rational>(0);
+                Node one = d_nm->mkConst<Rational>(1);
+                d_rangeAssertions.insert(d_nm->mkNode(kind::LEQ, zero, sigma));
+                d_rangeAssertions.insert(d_nm->mkNode(kind::LEQ, sigma, one));
```
@ahmed-irfan ?

ahmed-irfan replied:
I think keeping in the form of conjunctions would be better because it delays the integral reasoning. 
this will first go to the real solver first and if integral reasoning is required, the integer solver will (lazily) add the branch-and-bound clauses. 
Moreover, as @yoni206 said, this appears more consistent with the multiplication case.

yoni206 commented on src/preprocessing/passes/bv_to_int.cpp:
```diff
@@ -0,0 +1,1003 @@
+            {
+              Unimplemented();
+              break;
+            }
```
I do not eliminate them. I think that the right place to eliminate them is in `bv_to_bool` preprocessing pass.

mpreiner replied:
They will be introduced in the bool-to-bv pass.

yoni206 replied:
Now not allowing to turn on that one.

aniemetz replied:
The this should not be `Unimplemented()`, but you should assert before the switch statement that they do not occur.

yoni206 commented on src/preprocessing/passes/bv_to_int.h:
```diff
@@ -0,0 +1,255 @@
+   *   to the original bit-wise operation.
+   *
+   * For example: Suppose bvsize is 4, granularity is 1, and f(x,y) = x && y.
+   * Denote by ITE(a,b) the term: ite(a==0, ite(b==0, 0, 0), ite(b==1, 1, 0)).
```
I wanted it to be consistent with the next example, in which the granularity is 2 (and then perhaps ITE is a good name)?

yoni206 commented on src/preprocessing/passes/bv_to_int.h:
```diff
@@ -0,0 +1,255 @@
+   * In the translation to integers, it is convenient to assume that certain
+   * bit-vector operators do not occur in the original formula (e.g., repeat).
+   * This function eliminates all these operators.
+   */
```
I am not sure about maintainability of that list. Someone might decide some day to eliminate another operator, or to not eliminate an existing operator, and then there is danger they will forget to update the list.

In addition, this list is pretty clear in the big switch here, no?
https://github.com/yoni206/CVC4/blob/aebdbd4499ac8047f0c1334fc00b3208594ca04e/src/preprocessing/passes/bv_to_int.cpp#L242

yoni206 commented on test/regress/regress0/bv/bv_to_int_bitwise.smt2:
```diff
@@ -0,0 +1,8 @@
+; COMMAND-LINE: --solve-bv-as-int=1 --no-check-models  --no-check-unsat-cores --no-check-proofs
```
Then it should be set to 4 inside the pass. I am adding this to the test.

yoni206 commented on test/regress/regress0/bv/bvuf_to_intuf_smtlib.smt2:
```diff
@@ -0,0 +1,123 @@
+(set-info :category "industrial")
+
+
+
```
I can remember that for the next time, but for now it would be difficult to reproduce the bug.

yoni206 replied:
Will remember for next time. Right now it is difficult to reproduce that bug, I am not even sure which commit it was.

yoni206 commented on test/regress/regress0/bv/bv_to_int1.smt2:
```diff
@@ -0,0 +1,13 @@
+; COMMAND-LINE: --solve-bv-as-int=1 --no-check-models  --no-check-unsat-cores --no-check-proofs 
+; COMMAND-LINE: --solve-bv-as-int=2 --no-check-models  --no-check-unsat-cores --no-check-proofs
+; COMMAND-LINE: --solve-bv-as-int=4 --no-check-models  --no-check-unsat-cores --no-check-proofs
```
Added.

yoni206 commented on src/preprocessing/passes/bv_to_int.cpp:
```diff
@@ -0,0 +1,1003 @@
+    {
+      granularity = granularity - 1;
+    }
+  }
```
done.

yoni206 commented:
Thank you for this in-depth review!
I believe I addressed all your comments.
Please let me know if I did not, and/or if you have more.

mpreiner commented on src/preprocessing/passes/bv_to_int.h:
```diff
@@ -4,52 +4,53 @@
  ** Tr(c) = the integer value of c, for any bit-vector constant c.
- ** Tr((bvadd s t)) = Tr(s) + Tr(t) mod 2^k, where k is the bit-width of s and t.
+ ** Tr((bvadd s t)) = Tr(s) + Tr(t) mod 2^k, where k is the bit width of s and
+ *t.
```
Try to make a different line break here, this looks strange (add missing `* `).

mpreiner commented on src/smt/smt_engine.cpp:
```diff
@@ -3341,14 +3349,29 @@ void SmtEnginePrivate::processAssertions() {
+    {
+      /**
+       * The granularity sets the size of the ITE in each element
+       * of the sum that is generated for bit-wise operators.
```
bitwise

mpreiner commented on src/preprocessing/passes/bv_to_int.cpp:
```diff
@@ -330,53 +315,40 @@ Node BVToInt::bvToInt(Node n)
-                throw TypeCheckingException(
-                    current.toExpr(),
-                    string("Cannot translate const: ") + current.toString());
+              // Bit-vector cnostants are transformed into their integer value.
```
constants

mpreiner commented on src/preprocessing/passes/bv_to_int.cpp:
```diff
@@ -867,33 +842,28 @@ Node BVToInt::createShiftNode(vector<Node> children,
+    builder << d_nm->mkNode(kind::INTS_MODULUS_TOTAL,
+                            d_nm->mkNode(kind::INTS_DIVISION_TOTAL, x, ite),
+                            pow2(bvsize));
   }
```
Simpler:

```
kind::Kind then_kind = isLeftShift ? kind::MULT : kind::INTS_DIVISION_TOTAL;
return d_nm->mkNode(kind::ITE,
                              d_nm->mkNode(kind::LT, y, d_nm->mkConst<Rational>(bvsize)),
                              d_nm->mkNode(kind::INTS_MODULUS_TOTAL,
                                                            d_nm->mkNode(then_kind, x, ite),
                                                            pow2(bvsize)),
                              d_zero);
```

mpreiner commented on src/preprocessing/passes/bv_to_int.cpp:
```diff
@@ -947,16 +925,16 @@ Node BVToInt::createBitwiseNode(Node x,
   // f is defined over 1 bit, while the table is defined over `granularity` bits
   std::map<std::pair<uint64_t, uint64_t>, uint64_t> table;
-  for (uint64_t i = 0; i < ((uint64_t) pow(2, granularity)); i++)
+  for (uint64_t i = 0; i < ((uint64_t)pow(2, granularity)); i++)
```
Compute `((uint64_t)pow(2, granularity))` only once and assert table size.

mpreiner commented on src/preprocessing/passes/bv_to_int.cpp:
```diff
@@ -967,12 +945,21 @@ Node BVToInt::createBitwiseNode(Node x,
-  // create the big sum
+  /*
+   * create the sum
+   * For granulaity 1, the sum has bvsize elements.
```
granularity

mpreiner commented on src/preprocessing/passes/bv_to_int.cpp:
```diff
@@ -967,12 +945,21 @@ Node BVToInt::createBitwiseNode(Node x,
+   * In contrast, if bvsize = granularity, sum has one element.
+   * Each element in the sum is an ite that corresponds to the generated table,
+   * multiplied by the appropriate power of two.
+   * More details are in the .h file.
```
bv_to_int.h

yoni206 commented on src/preprocessing/passes/bv_to_int.cpp:
```diff
@@ -0,0 +1,1003 @@
+
+// wrapper function to cmath's pow function.
+// Casts the result to an integer.
+Rational intpow2(uint64_t b) { 
```
now i am using it inside this function.

yoni206 commented on src/preprocessing/passes/bv_to_int.cpp:
```diff
@@ -0,0 +1,1003 @@
+{
+  Assert(granularity <= 8);
+  Node ite = d_nm->mkConst<Rational>(table[std::make_pair(0, 0)]);
+  for (uint64_t i = 0; i < ((uint64_t) pow(2, granularity)); i++)
```
done.

mpreiner commented on src/preprocessing/passes/bv_to_int.cpp:
```diff
@@ -0,0 +1,982 @@
+      for (uint64_t n = 0; n < granularity; n++)
+      {
+        // b is the result of f on the current bit
+        bool b = f(((((uint64_t)(i / pow(2, n))) % 2) == 1),
```
```suggestion
        bool b = f((i >> n) & 1 == 1),
```

mpreiner commented on src/preprocessing/passes/bv_to_int.cpp:
```diff
@@ -0,0 +1,982 @@
+      {
+        // b is the result of f on the current bit
+        bool b = f(((((uint64_t)(i / pow(2, n))) % 2) == 1),
+                   (((j / ((uint64_t)pow(2, n))) % 2) == 1));
```
```suggestion
                   (j >> n) & 1 == 1));
```

mpreiner commented on src/preprocessing/passes/bv_to_int.cpp:
```diff
@@ -0,0 +1,982 @@
+        // add the corresponding power of 2 only if the result is 1
+        if (b)
+        {
+          sum += pow(2, n);
```
```suggestion
          sum += 1 << n;
```

aniemetz commented on src/smt/smt_engine.cpp:
```diff
@@ -3337,6 +3349,31 @@ void SmtEnginePrivate::processAssertions() {
+    {
+      throw ModalException(
+          "solving bitvectors as integers is currently not supported "
+          "incrementally.");
```
```suggestion
          "when solving incrementally.");
```

aniemetz commented on src/smt/smt_engine.cpp:
```diff
@@ -1406,6 +1413,11 @@ void SmtEngine::setDefaults() {
 
+    if (options::solveBVAsInt() > 0)
+    {
+      options::bitvectorToBool.set(true);
```
Why do you enable bv-to-bool when bv-to-int is enabled?

aniemetz commented on src/smt/smt_engine.cpp:
```diff
@@ -3337,6 +3349,31 @@ void SmtEnginePrivate::processAssertions() {
+       * Since we don't want to introduce ITEs with unbounded size,
+       * we bound the granularity.
+       */
+      throw ModalException("solve-bv-as-int accepts values from 0 to 8.");
```
Is 8 the maximum size we can handle?

mpreiner replied:
Note: This does not have anything to do with bit-vector size. It's the partition size in which the bit-vector is partitioned.

yoni206 replied:
@mpreiner : correct.
@aniemetz : We haven't tried higher than 8, but this already gives ITEs of size `2^16`.

aniemetz commented on src/preprocessing/passes/bv_to_int.h:
```diff
@@ -0,0 +1,263 @@
+ ** The size of each block is the granularity, and so the number of
+ ** blocks is:
+ ** bit width/granularity (rounded down).
+ ** We create an ITE that resresents an arbitrary block,
```
```suggestion
 ** We create an ITE that represents an arbitrary block,
```

aniemetz commented on src/preprocessing/passes/bv_to_int.h:
```diff
@@ -0,0 +1,263 @@
+ ** appropriate power of two.
+ ** More formally:
+ ** Let g denote the granularity.
+ ** Let k denote the bit width of s and t.
```
In smt_engine.cpp, you use `k` for the granularity -- might be better to use consistent names?

yoni206 replied:
changed in smt_engine.

aniemetz commented on src/preprocessing/passes/bv_to_int.h:
```diff
@@ -0,0 +1,263 @@
+   * 3 | 2 | 2
+   * 3 | 3 | 3
+   *
+   * (for example, 2 in binary is 10 and 1 in binary is 01, and so doing
```
```suggestion
   * For example, 2 in binary is 10 and 1 in binary is 01, and so doing
```

aniemetz commented on src/preprocessing/passes/bv_to_int.h:
```diff
@@ -0,0 +1,263 @@
+   * 3 | 3 | 3
+   *
+   * (for example, 2 in binary is 10 and 1 in binary is 01, and so doing
+   * "bitwise f" on them gives 00).
```
```suggestion
   * "bitwise f" on them gives 00.
```

aniemetz commented on src/preprocessing/passes/bv_to_int.h:
```diff
@@ -0,0 +1,263 @@
+  /**
+   * A helper function for createBitwiseNode
+   * - x and y are integer nodes that correspond to the original bit-vector
+   * nodes.
```
```suggestion
   *   nodes.
```

aniemetz commented on src/preprocessing/passes/bv_to_int.h:
```diff
@@ -0,0 +1,263 @@
+
+  /**
+   * A helper function for createBitwiseNode
+   * - x and y are integer nodes that correspond to the original bit-vector
```
```suggestion
   * - x, y: integer nodes that correspond to the original bit-vector
```

aniemetz commented on src/preprocessing/passes/bv_to_int.h:
```diff
@@ -0,0 +1,263 @@
+   * A helper function for createBitwiseNode
+   * - x and y are integer nodes that correspond to the original bit-vector
+   * nodes.
+   * - granularity represents the bitwidth of the original bit-vector nodes.
```
```suggestion
   * - granularity: the bitwidth of the original bit-vector nodes.
```

aniemetz commented on src/preprocessing/passes/bv_to_int.h:
```diff
@@ -0,0 +1,263 @@
+   * - x and y are integer nodes that correspond to the original bit-vector
+   * nodes.
+   * - granularity represents the bitwidth of the original bit-vector nodes.
+   * - table represents a function from pairs of integers to integers.
```
```suggestion
   * - table: a function from pairs of integers to integers.
```

aniemetz commented on src/preprocessing/passes/bv_to_int.h:
```diff
@@ -0,0 +1,263 @@
+   * the result is 0.
+   * - children: the two operands for the shift
+   * - bvsize: the original bit widths of the operands
+   *   (before translation to integers)
```
```suggestion
   *           (before translation to integers)
```

aniemetz commented on src/preprocessing/passes/bv_to_int.h:
```diff
@@ -0,0 +1,263 @@
+  /**
+   * n is a bit-vector term or formula.
+   * The result is an integer term and is computed
+   * according to the translation specified above.
```
Move first line to the bottom, first description of the function, then description of the parameters.

aniemetz commented on src/preprocessing/passes/bv_to_int.h:
```diff
@@ -0,0 +1,263 @@
+   * variable, we need to guard the range of the newly introduced variable.
+   * For bit width k, the constraint is 0 <= newVar < 2^k.
+   * - newVar is the newly introduced integer variable
+   * - k is the bit width of the original bit-vector variable.
```
Try to use consistent formatting for how you describe the parameters, I like how you did it above with `param: description`.

aniemetz replied:
This applies to the whole file, will not mark every instance.

aniemetz commented on src/preprocessing/passes/bv_to_int.cpp:
```diff
@@ -0,0 +1,981 @@
+  Integer two = Integer(2);
+  Integer p = two.pow(b);
+  Rational r = Rational(p, one);
+  return r;
```
You could make a one liner out of this:
`return Rational(Integer(2).pow(b), Integer(1));`
Less verbose, IMO a bit faster to comprehend, but that's of course personal preference.

yoni206 replied:
done.

aniemetz commented on src/preprocessing/passes/bv_to_int.cpp:
```diff
@@ -0,0 +1,981 @@
+using namespace CVC4::theory;
+using namespace CVC4::theory::bv;
+
+Rational intpow2(uint64_t b)
```
You should wrap all helper functions (= that are not members of the preprocessing pass class)in an unnamed namespace (see https://github.com/CVC4/CVC4/wiki/Code-Style-Guidelines#unnamed-namespaces-and-static-variables):
```
namespace {
...
}
```

yoni206 replied:
done.

aniemetz commented on src/preprocessing/passes/bv_to_int.cpp:
```diff
@@ -0,0 +1,981 @@
+          }
+          else
+          {
+            Unimplemented();
```
Is this supposed to be a TODO? Or a case that can never happen?
If the latter, you should assert at the top that `current.isVar() || current.isConst()` and case split between those two. If it is a TODO, why is it not handled yet?

yoni206 replied:
The code now has only a single `Unimplemented()`, due to lack of support for logics that go beyond `QF_UFBV`, and this is now documented in the code. For this particular case that you commented on, I replaced it with the assertion that you suggested.

aniemetz commented on src/preprocessing/passes/bv_to_int.cpp:
```diff
@@ -0,0 +1,981 @@
+          vector<Node> intized_children;
+          for (uint64_t i = 0; i < currentNumChildren; i++)
+          {
+            intized_children.push_back(d_bvToIntCache[current[i]]);
```
I personally find `intized_children` to be a bit of a weird name choice, made me double back while reading through the code and was not immediately obvious what it is supposed to mean. Maybe use a better name?

yoni206 replied:
I find-replaced it to `translated_children`.

aniemetz commented on src/preprocessing/passes/bv_to_int.cpp:
```diff
@@ -0,0 +1,981 @@
+            }
+            case kind::BITVECTOR_NEG:
+            {
+              // we use an ITE for the case where the operand is 0.
```
Why is this case split necessary here?

yoni206 replied:
This is an optimization to avoid introducing a `mod`. I now document it in the code.

aniemetz commented on src/preprocessing/passes/bv_to_int.cpp:
```diff
@@ -0,0 +1,981 @@
+   Assert(table.size() == max_value * max_value);
+
+  /*
+   * create the sum
```
```suggestion
   * Create the sum.
```

yoni206 replied:
fixed.

yoni206 commented:
Thanks for these additional comments, @aniemetz. I believe I have addressed all your comments now.
In addition, @mpreiner, I also fixed a problem that occurred in the elimination pass. I used the two maps in a wrong way and found out about it by testing with `QF_BV` benchmarks from `SMT-LIB`. Now this is fixed.
